### DISCUSSION TITLE
MaterialGrid interpolation for two LorentzianSusceptibility materials

### DISCUSSION THREAD

**Discussion by yungchiang** [NanoComp/meep#2006] [Q&A]
Dear meep community,

I try to learn the adjoint simulation of optimization of meep. When I use the tutorial examples of adjoint solver, everything is fine. The same results can be obtained. However, we I try to use the MaterialGrid interpolation for two LorentzianSusceptibility materials, I found that only the epsilon part can be interpolated. The E_susceptibilities part is completely ignored. Therefore, the interpolated material is only a dielectric material not a LorentzianSusceptibility material. My simplified code for testing MaterialGrid is listed below. Can anyone help me to find what is wrong? The meep version I used is 1.20. Thank you for your help.

```
import meep as mp
import numpy as np
from matplotlib import pyplot as plt
from matplotlib.patches import Circle
from scipy import special, signal

Si = mp.Medium(index=3.4)
SiO2 = mp.Medium(index=1.44)
air = mp.Medium(index=1.0)
#
aa = 1.0e-6
cc = 3.0e8
wp1 = 1.4e16
rd1 = 1.5e14
wwp1 = wp1/(2.0*np.pi)*(aa/cc)
rrd1 = rd1/(2.0*np.pi)*(aa/cc)
www0 = wwp1/100.0
sigma1 = wwp1/www0
wp2 = 2.5e16
rd2 = 1.2e14
wwp2 = wp2/(2.0*np.pi)*(aa/cc)
rrd2 = rd2/(2.0*np.pi)*(aa/cc)
sigma2 = wwp2/www0
avg_sig = 0.5*(sigma1*sigma1+sigma2*sigma2)
Ag_susc1=[mp.LorentzianSusceptibility(frequency=www0,gamma=rrd1,sigma=sigma1*sigma1)]
Ag_my=mp.Medium(epsilon=5.0, E_susceptibilities=Ag_susc1)
Au_susc1=[mp.LorentzianSusceptibility(frequency=www0,gamma=rrd1,sigma=sigma2*sigma2)]
Au_my=mp.Medium(epsilon=3.9, E_susceptibilities=Au_susc1)
Avg_susc1=[mp.LorentzianSusceptibility(frequency=www0,gamma=rrd1,sigma=avg_sig)]
Avg_my=mp.Medium(epsilon=4.475, E_susceptibilities=Avg_susc1)
#
waveguide_width = 0.5
design_region_width = 1.0
design_region_height = 1.0
waveguide_length = 0.5
pml_size = 0.4
resolution = 30
frequencies = 1/np.linspace(1.5,1.6,10)
#
design_region_resolution = int(1*resolution)
#
Sx = 2*pml_size + 6*waveguide_length + design_region_width
Sy = 2*pml_size + 5*waveguide_length + design_region_height
cell_size = mp.Vector3(Sx,Sy)

pml_layers = [mp.PML(pml_size)]

fcen = 1/1.55
width = 0.2
fwidth = width * fcen
source_center  = [-Sx/2+pml_size+0.1,0,0]
source_size    = mp.Vector3(0,Sy,0)
kpoint = mp.Vector3(1,0,0)
src = mp.ContinuousSource(frequency=1/1.55,fwidth=fwidth)
source = [mp.EigenModeSource(src,
                    eig_band = 1,
                    direction=mp.NO_DIRECTION,
                    eig_kpoint=kpoint,
                    eig_parity=mp.EVEN_Z,
                    size = source_size,
                    center=source_center)]
#----------------------------------------------------------------
Nx = int(design_region_resolution*design_region_width)
Ny = int(design_region_resolution*design_region_height)
x = np.ones((Nx,Ny))-0.5
design_variables = mp.MaterialGrid(mp.Vector3(Nx,Ny),
mp.Medium(epsilon=5.0,E_susceptibilities=Ag_susc1),
            mp.Medium(epsilon=5.0,E_susceptibilities=Au_susc1),
            weights=x, grid_type='U_DEFAULT')
geometry = [mp.Block(center=mp.Vector3(), size=mp.Vector3(design_region_width,design_region_height,0), material=design_variables), 
 ]
#----------------------------------------------------------------
sim = mp.Simulation(cell_size=cell_size,
           boundary_layers=pml_layers,
           geometry=geometry,
           sources=source,
           default_material=air,
           resolution=resolution)
sim.plot2D()
plt.show()

sim.run(until=200)
plt.figure()
sim.plot2D(fields=mp.Ez)
plt.show()
plt.figure()
sim.plot2D(output_plane=mp.Volume(center=(0,0,0),
size=(Sx-0.4,Sy-0.4,0)),fields=mp.Ex)
plt.show()
plt.figure()
sim.plot2D(output_plane=mp.Volume(center=(0,0,0),
size=(Sx-0.4,Sy-0.4,0)),fields=mp.Ey)
plt.show()
plt.figure()
sim.plot2D(output_plane=mp.Volume(center=(0,0,0),
size=(Sx-0.4,Sy-0.4,0)),fields=mp.Hz)
plt.show()
```

**Comment by smartalecH** [comment]
> However, we I try to use the MaterialGrid interpolation for two LorentzianSusceptibility materials, I found that only the epsilon part can be interpolated. The E_susceptibilities part is completely ignored.

How did you determine this? Did you use `plot2D` and explicitly look at the geometry at two different frequency points using the `frequency` parameter? (see the docstring [here](https://meep.readthedocs.io/en/latest/Python_User_Interface/#data-visualization))

[Here's](https://github.com/NanoComp/meep/discussions/1997#discussioncomment-2342060) a discussion describing how dispersive materials are interpolated.

**Reply by yungchiang** [reply]
I use the continuous source (frequency=1/1.55, fwidth=fwidth) in the simulation and observe the simulated Hz, Ex and Ey fields. I also do another simulation with a manually interpolated LorentzianSusceptibility materal and also observe the fields.
The geometry is reset to the new manual interpolated material listed below:
geometry = [
   mp.Block(center=mp.Vector3(),
   size=mp.Vector3(design_region_width,design_region_height,0), 
    material=Avg_my), # design region   
]
The simulated fields show that the material of the structure is only a dielectric in the first case but a LorentzianSusceptibility materal in the second case.

**Comment by yungchiang** [comment]
Hello, Meep community,

For MaterialGrid of LorentzianSusceptibility material, I still can’t find what is wrong in my setting. I do another three simulations to test it. In this simulation, a square structure is put in the center of the simulation domain. In the first test, the material of the square is set as MaterialGrid with medium1 and medium2 to be the same LorentzianSusceptibility material. In the second test, the material of the square is set to LorentzianSusceptibility material of medium 1. In the third simulation, the material is set a dielectric with the epsilon value equal to that of medium 1. In the simulation, a continuous source (TM mode) is incident onto the square. The simulation results show that the first and third tests are the same (i.e. the square is a dielectric). However, in the second case, the square is a metal (LorentzianSusceptibility material). The python codes are listed below. Do I misunderstand the usage of MaterialGrid for LorentzianSusceptibility material? How do I correct the code? Can anyone help me?
#-------------------------------------------------------------------------
import meep as mp
import numpy as np
from matplotlib import pyplot as plt
from matplotlib.patches import Circle
from scipy import special, signal

air = mp.Medium(epsilon=1.0)
die_my = mp.Medium(epsilon=5.0)

aa = 1.0e-6
cc = 3.0e8
wp1 = 1.4e16
rd1 = 1.5e14
wwp1 = wp1/(2.0np.pi)(aa/cc)
rrd1 = rd1/(2.0np.pi)(aa/cc)
www0 = wwp1/100.0
sigma1 = wwp1/www0
Ag_susc1=[mp.LorentzianSusceptibility(frequency=www0,gamma=rrd1,sigma=sigma1*sigma1)]
Ag_my=mp.Medium(epsilon=5.0, E_susceptibilities=Ag_susc1)

design_region_width = 1.0
design_region_height = 1.0
waveguide_length = 0.5
pml_size = 0.4
resolution = 30
frequencies = 1/np.linspace(1.5,1.6,10)

design_region_resolution = int(1*resolution)

Sx = 2pml_size + 6waveguide_length + design_region_width
Sy = 2pml_size + 5waveguide_length + design_region_height
cell_size = mp.Vector3(Sx,Sy)
pml_layers = [mp.PML(pml_size)]
fcen = 1/1.55
width = 0.2
fwidth = width * fcen
source_center = [-Sx/2+pml_size+0.1,0,0]
source_size = mp.Vector3(0,Sy,0)
kpoint = mp.Vector3(1,0,0)
src = mp.ContinuousSource(frequency=1/1.55,fwidth=fwidth)
source = [mp.EigenModeSource(src,
eig_band = 1,
direction=mp.NO_DIRECTION,
eig_kpoint=kpoint,
eig_parity=mp.EVEN_Z,
size = source_size,
center=source_center)]
#-----------------------------------------------------------------------
Nx = int(design_region_resolutiondesign_region_width)
Ny = int(design_region_resolutiondesign_region_height)
x = np.zeros((Nx,Ny))+0.5
design_variables = mp.MaterialGrid(mp.Vector3(Nx,Ny),Ag_my,Ag_my,weights=x,do_averaging=True,grid_type='U_DEFAULT')
geometry = [
mp.Block(center=mp.Vector3(),size=mp.Vector3(design_region_width,design_region_height,0),
                material=design_variables),
                # material=Ag_my),
                # material=die_my),
]
#----------------------------------------------------------------------
sim = mp.Simulation(cell_size=cell_size,
boundary_layers=pml_layers,
geometry=geometry,
sources=source,
default_material=air,
resolution=resolution)
sim.plot2D(frequency=1/1.55)
plt.show()
vals_eps = []
def get_eps_slice(sim):
vals_eps.append(sim.get_epsilon_point(mp.Vector3(0,0,0),frequency=1/1.55))
sim.run(mp.at_every(100,get_eps_slice),until=200)
plt.figure()
sim.plot2D(fields=mp.Ez)
plt.show()
plt.figure()
sim.plot2D(fields=mp.Ex,plot_eps_flag=False)
plt.show()
plt.figure()
sim.plot2D(fields=mp.Ey,plot_eps_flag=False)
plt.show()
plt.figure()
sim.plot2D(fields=mp.Hz,plot_eps_flag=False)
plt.show()

**Comment by smartalecH** [comment]
Duplicated and answered in #2037.
